### DISCUSSION TITLE
Deploy Helm Chart with additional resources in same kustomization

### DISCUSSION THREAD

**Discussion by nicogl** [fluxcd/flux2#1770] [Q&A] [answered]
I have this kustomization.yaml file that specifies a HelmRelease and a Namespace:

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: flux-system
resources:
  - my-helm-release.yaml
  - namespace.yaml
```

The Helm Chart pointed by the HelmRelease doesn't create any namespace, therefore I added the namespace.yaml to create the Namespace myself.
The idea is that the previous Helm Chart installs its resources inside the mentioned namespace. I did this by specifying the _**targetNamespace**_ property in the HelmRelease file.

The problem is that the Namespace is not being created.

What am I doing wrong? Is there any way to install a Helm Chart along with other kubernetes resources as part of the same kustomization?

**Comment by stefanprodan** [comment, marked as answer]
You are overriding the namespace to flux-system, remove it from your kustomization.yaml.

**Reply by nicogl** [reply]
@stefanprodan I removed the entire line but I'm afraid it doesn't work...

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - my-helm-release.yaml
  - namespace.yaml
```

I get below error:

_helmrelease/my-helm-release.flux-system
Helm install failed: namespaces "my-namespace" not found_

I would assume that the Namespace object was created before than the HelmRelease... Is there any way to specify a creation order? Note that the Namespace is not part of a HelmRelease, but an isolated resource declared in a separate file, so there's no chance to set a **`dependsOn`** property on the HelmRelease.

**Reply by nicogl** [reply]
@stefanprodan I think I got the solution for this issue.
The problem was that the Helm Chart doesn't create a namespace in which to install its resources, and that's why I had to include the namespace.yaml file...

The solution then was to use the `**spec.install.createNamespace**` property of the HelmRelease with `true` value.
Now the Namespace gets created before the HelmRelease deploy runs.

I'm still wondering though if it's really possible to have the namespace.yaml file pointed by the kustomization and create the Namespace that way...

Thanks anyway!

**Reply by stefanprodan** [reply]
kustomize-controller does create namespaces before anything else. When this kustomization is applied on the cluster, what does kustomize-controller logs say, do you see the namespace being applied first?

**Reply by nicogl** [reply]
I'm afraid the kustomize-controller never refers to the namespace... either with error or failure.
If I get the HelmReleases I see the message `install retries exhausted` for the release because it doesn't find the namespace. Along with the alert messages in Slack: _**reconciliation failed: Helm install failed: namespaces "my-namespace" not found**_ and _**reconciliation failed: install retries exhausted**_.

Keep in mind that the Kustomization file is a `kustomize.config.k8s.io/v1beta1` object, not a Flux Kustomization like: `kustomize.toolkit.fluxcd.io/v1beta1`

As I said before I could sort this out because it's a namespace object so I could use the HelmRelease `spec.install.createNamespace` property, but if I wanted to create a different type of resource I guess the issue would remain and I could not create it as part of the kustomization.

I can try any suggestion you may have in order to see if this is really possible, thanks!
